conformance/tdl/variable_expansion.ion

```diff
@@ -3,15 +3,125 @@
+               (mactab (macro foo (x) (% x null)))
+               (mactab (macro foo (x) (% x %)))
+               (mactab (macro foo (x) (% x ?)))
+               (mactab (macro foo (x) (% x +)))
+               (mactab (macro foo (x) (% x *)))
+               "must correspond to a valid, in-scope variable binding"
+               (mactab (macro foo (x) (%y)))
+               // `foo` cannot access `y` from where it is invoked in `bar`
+               (mactab (macro foo (x) (%y))
+                       (macro bar (y) (.foo 1)))
+               (signals "invalid macro definition")))
+
+// EVALUATION TEST CASES
+
+(ion_1_1 "when expanding"
+         (then "a zero-or-one parameter"
+               (mactab (macro foo (x?) (%x)))
+               (then "with zero arguments"
+                     (toplevel ('#$:foo' ('#$::')))
+                     (produces))
+               (then "with one argument"
+                     (toplevel ('#$:foo' ('#$::' 1)))
+                     (produces 1))
+               (then "with more than one argument"
+                     (toplevel ('#$:foo' ('#$::' 1 2 3)))
+                     (signals "too many arguments")))
+         (then "an exactly-one parameter"
+               (mactab (macro foo (x?) (%x)))
+               (then "with zero arguments"
+                     (toplevel ('#$:foo' ('#$::')))
+                     (signals "too few arguments"))
+               (then "with one argument"
+                     (toplevel ('#$:foo' ('#$::' 1)))
+                     (produces 1))
+               (then "with more than one argument"
+                     (toplevel ('#$:foo' ('#$::' 1 2 3)))
+                     (signals "too many arguments")))
+         (then "a one-or-more parameter"
+               (mactab (macro foo (x+) (%x)))
+               (then "with zero arguments"
+                     (toplevel ('#$:foo' ('#$::')))
+                     (signals "too few arguments"))
+               (then "with one argument"
+                     (toplevel ('#$:foo' ('#$::' 1)))
+                     (produces 1))
+               (then "with more than one argument"
+                     (toplevel ('#$:foo' ('#$::' 1 2 3)))
+                     (produces 1 2 3)))
+         (then "a zero-or-more parameter"
+               (mactab (macro foo (x*) (%x)))
+               (then "with zero arguments"
+                     (toplevel ('#$:foo' ('#$::')))
+                     (produces))
+               (then "with one argument"
+                     (toplevel ('#$:foo' ('#$::' 1)))
+                     (produces 1))
+               (then "with more than one argument"
+                     (toplevel ('#$:foo' ('#$::' 1 2 3)))
+                     (produces 1 2 3))))
+
+(ion_1_1 "the argument arity should not be checked for a variable that unused"
+         (then "not enough args for exactly-one"
+               (mactab (macro foo (x) 123))
+               (toplevel ('#$:foo'))
+               (produces 123))
```

Comment:
Why allow these?

Reply:
We need to validate the number of arguments in the call in order for the binary encoding to be readable. There's no way to invoke `foo` above in binary without providing an expression for `x`.

Reply:
Oh, right. This is an "exactly one" use case so it would be caught by e-expression parser. Okay. What if we had something like this, what should it do?
```
(then "not enough args for exactly-one"
               (mactab (macro foo (x) 123)
                              (macro bar () (.foo (.none))) )
               (toplevel ('#$:bar'))
               (produces 123))
```

Reply:
The call to `(.foo (.none))` is syntactically valid, but would fail during evaluation. (A compiler could notice this, but i you replace this `.none` with a call to some other macro whose expansion is unknown, the problem stands.)

The only place you're guaranteed to be able to report the error is during the expansion of `x`. In a lazy evaluator, this would not occur. In an eager one, it would. I think of it as 'skipped' data in that it's wrong but may never be evaluated.

Reply:
I've been thinking about this because something doesn't quite sit right with me about it.

> In a lazy evaluator, this would not occur. In an eager one, it would.

The implication is that an Ion data stream could be readable by a lazy implementation but not by an eager implementation. I _think_ we need to ensure that every spec-compliant implementation is capable of reading the same Ion data streams as every other spec-compliant implementation. (Limitations on the max size of values notwithstanding.)  I _think_ we need to specify it one way or the other.